crates/virtual-node/src/lib.rs

```diff
@@ -101,54 +108,99 @@ impl VirtualNode {
     ///
     /// let div = VirtualNode::text("div");
     /// ```
-    pub fn text(text: &str) -> VirtualNode {
-        VirtualNode {
-            tag: "".to_string(),
-            props: HashMap::new(),
-            events: Events(HashMap::new()),
-            children: Some(vec![]),
-            text: Some(text.to_string()),
-        }
+    pub fn text(text: &str) -> Self {
+        VirtualNode::Text(text.into())
     }
-}
 
-/// A web_sys::Element along with all of the closures that were created for that element's
-/// events and all of it's child element's events.
-pub struct CreatedElement {
-    /// An Element that was created from a VirtualNode
-    pub element: Element,
-    /// A map of an element's unique identifier along with all of the Closures for that element.
+    /// Create a new `VirtualDomElement`.
     ///
-    /// The DomUpdater uses this to look up elements and see if they're still in the page. If not
-    /// the refernce that we maintain to their closure will be dropped, thus freeing the Closure's
-    /// memory.
-    pub closures: HashMap<u32, Vec<DynClosure>>,
-}
+    /// This is the variant that is contained inside `VirtualNode::Element`. It
+    /// can be turned into a `VirtualNode` using the from/into.
+    pub fn element_variant<S: Into<String>>(tag: S) -> VirtualNodeElement {
+        VirtualNodeElement::new(tag)
+    }
 
-impl Deref for CreatedElement {
-    type Target = Element;
+    /// Create a new `VirtualDomText`.
+    ///
+    /// This is the variant that is contained inside `VirtualNode::Text`. It
+    /// can be turned into a `VirtualNode` using the from/into.
+    pub fn text_variant<S: Into<String>>(text: S) -> VirtualNodeText {
```

Comment:
I think that these `variant` methods feel a bit odd. I'd almost prefer:

```rust
VirtualNode::text("some text").as_text().unwrap()
```

In order to access the underlying `VirtualTextNode`.

That way you aren't creating these subtypes directly, but instead only creating `VirtualNode`'s and then working through that APi. It feels like that smaller surface area would still get the job done just fine but with fewer methods to worry about. Especially since ergonomics aren't very important since people shouldn't be creating `VirtualNode`'s by hand and should be using a macro.

Is this tremendously different? No not really.. But something about needing to make a variant and then turn it into a node feels a little off vs. making a `VirtualNode` and then working with that `VirtualNode`.

Thoughts?

Reply:
I'm a bit undecided on this.

First off: Why would you recommend to always create nodes through the `html` macro? Isn't `let div = html! { <div> </div> }` both slower and potentially more error-prone (due to the HTML parsing) than `let div = VirtualNode::element("div")`?

I love the html macro for building up a tree of virtual nodes or for a single node with attributes, but for single nodes without attributes I'd personally prefer a constructor method.

However, in many cases users will be creating virtual nodes programmatically. This is the case in a project of mine for example. I have some internal state, then generate `VirtualNode` elements from that. On state changes, those elements are diffed against the previous state, and patches are applied to the actual DOM.

When generating virtual element nodes directly / programmatically, I would prefer something like this:

```rust
let vnode = VElement::builder(element_type)
    .attr("id", "wrapper")
    .attr("class", "myclass")
    .append_child(VirtualNode::text("foobar"))
    .build()
    .into::<VirtualNode>();
```

...over this:

```rust
let mut vnode = VirtualNode::element(element_type);
vnode.as_element_mut().unwrap().props.insert("id", "wrapper");
vnode.as_element_mut().unwrap().props.insert("class", "myclass");
vnode.as_element_mut().unwrap().children = vec![VirtualNode::text("foobar")];
```

I'm not a fan of the unwraps, even though in this case we can be sure that the unwrapping cannot fail. But the type system does not know it. And if we later change `VirtualNode::element` to `VirtualNode::text` the compiler will not complain, but we'll get a runtime exception.

What are your thoughts? I agree that the current API of `VElement` and `VText` aren't great btw! But I think we could have a nice API for those variants, without trying to hide their constructors.

Reply:
> However, in many cases users will be creating virtual nodes programmatically. This is the case in a project of mine for example.

Ahhh alright in my mind I couldn't think of realistic scenarios where people would be building nodes with methods instead of just using the macro but if you're already doing this in the real world then that's already evidence that I was wrong!

Cool - let's keep is this way and just continue to evolve the API as we feel it out

Reply:
One concern that I have is people needing to know about virtual elements vs. virtual nodes vs. virtual text which I'd imagine can get confusing.

I'd imagine that every few people know off the top of their head that a Node is an upcast of an Element so the less you need to deal with that the better I think.

I think that requiring so many `.into()`s is less than ideal.

---

I really like the builder pattern that you illustrated though. Anyways - tl;dr I'm cool with going with whatever you think is best for now and then feeling it out a bit more.

Reply:
> One concern that I have is people needing to know about virtual elements vs. virtual nodes vs. virtual text which I'd imagine can get confusing.

Yep, I agree there. One way we could improve this would be putting the builder on the `VirtualNode` instead of the `VElement`. Then you can eliminate many cases where currently you need direct access to the `VElement`.

But that can be done later!